include/ibamr/WaveGenerationStrategy.h

```diff
@@ -0,0 +1,220 @@
+//      contributors may be used to endorse or promote products derived from
+//      this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+/////////////////////// INCLUDE GUARD ////////////////////////////////////
+
+#ifndef included_WaveGenerationStrategy
+#define included_WaveGenerationStrategy
+
+///////////////////////////// INCLUDES ///////////////////////////////////
+#include <string>
+
+#include "Variable.h"
+#include "ibamr/AdvDiffHierarchyIntegrator.h"
+#include "ibamr/INSVCStaggeredHierarchyIntegrator.h"
+#include "ibtk/ibtk_utilities.h"
+#include "tbox/Pointer.h"
+
+namespace IBAMR
+{
+/*!
+ * A collection of post processing call back functions to be hooked into
+ * IBAMR::HierarchyIntegrator class to employ a wave generation zone of some
+ * prescribed length within the computational domain. Additional strategies
+ * can be implemented as simple C functions within this collection.
+ *
+ * The instance of a WaveGenerationStrategy and the particular generation function
+ * should be registered with an appropriate hierarchy integrator via
+ * registerPostprocessIntegrateHierarchyCallback.
+ *
+ * \param ctx is the pointer to WaveGenerationStrategy struct.
+ */
+
+namespace WaveGenerationFunctions
+{
+void callStokesWaveRelaxationCallbackFunction(double current_time,
+                                              double new_time,
+                                              bool skip_synchronize_new_state_data,
+                                              int num_cycles,
+                                              void* ctx);
+} // namespace WaveGenerationFunctions
+
+/*! Struct for generating waves based on relaxation method.
+ *
+ */
+struct WaveGenerationStrategy
+{
+    /*
+     * Pointers to the fluid and advection-diffusion integrators.
+     */
+    SAMRAI::tbox::Pointer<IBAMR::INSVCStaggeredHierarchyIntegrator> d_ins_hier_integrator;
+    SAMRAI::tbox::Pointer<IBAMR::AdvDiffHierarchyIntegrator> d_adv_diff_hier_integrator;
+
+    /*
+     * Pointer to the level set variable representing the wave interface.
+     */
+    SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > d_phi_var;
```

Comment:
Also needs the `VariableContext`?

Reply:
It is algorithm dependent. I do this as a post-processing callback, and just use new_context directly.

Reply:
Should this class record the relevant contexts? Or if it should always be "new", could we put it in? I don't like the fact that users have to have extra knowledge about which context to use to actually access the data.

Reply:
It is not always "new". For example one can generate waves before the beginning of timestep instead of at the end. Former will use "current" context and latter will use new context. It is a timestepping choice. I just do it at the end of the timestep.

Reply:
One can also do both.

Reply:
I would suggest getting rid of this base class for now and merging the data members into the derived class.

Where are the functions that actually set the data? I think that they should tell the caller where to set the data values (either by passing in a patch data index or a (var,ctx) pair).

Reply:
Since it is a `struct` with public data members, I just set it in main.cpp directly. It is a lightweight structure utility that will be common for ALL wave theory implementations. Removing it will replicate these data members for all kinds of waves. Prefer doing that? Either way is fine by me.

Reply:
I would suggest to make a class like `WaveGenerationData` to store data members --- not to accomplish this through inheritance.

Reply:
Ah, yes, that sounds like the right design pattern.

Reply:
Looks good. Can you split into two headers --- one for `WaveGenerationData` and the other for `StokesWaveGenerator`?

If you wanted to indicate that the base class is an interface and not a concrete class, you could consider renaming `StokesWaveGenerator` to `StokesWaveGeneratorStrategy` since it is indeed specifying a generic interface for Stokes waves. OTOH, the current name is also fine.